src/index.ts

```diff
@@ -0,0 +1,257 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+import * as extend from 'extend';
+import * as util from 'util';
+import {Writable} from 'stream';
+const logging = require('@google-cloud/logging');
+// var Writable = require('stream').Writable;
+
+// Map of Stackdriver logging levels.
+const BUNYAN_TO_STACKDRIVER: {[key: number]: string} = {
+  60: 'CRITICAL',
+  50: 'ERROR',
+  40: 'WARNING',
+  30: 'INFO',
+  20: 'DEBUG',
+  10: 'DEBUG',
+};
+
+/*!
+ * Key to use in the Bunyan payload to allow users to indicate a trace for the
+ * request, and to store as an intermediate value on the log entry before it
+ * gets written to the Stackdriver logging API.
+ */
+export const LOGGING_TRACE_KEY = 'logging.googleapis.com/trace';
+
+/*!
+ * Gets the current fully qualified trace ID when available from the
+ * @google-cloud/trace-agent library in the LogEntry.trace field format of:
+ * "projects/[PROJECT-ID]/traces/[TRACE-ID]".
+ */
+function getCurrentTraceFromAgent() {
+  const agent = global._google_trace_agent;
+  if (!agent || !agent.getCurrentContextId || !agent.getWriterProjectId) {
+    return null;
+  }
+
+  const traceId = agent.getCurrentContextId();
+  if (!traceId) {
+    return null;
+  }
+
+  const traceProjectId = agent.getWriterProjectId();
+  if (!traceProjectId) {
+    return null;
+  }
+
+  return `projects/${traceProjectId}/traces/${traceId}`;
+}
+
+export class LoggingBunyan extends Writable {
+  private logName: string;
+  private resource: MonitoredResource|undefined;
+  private serviceContext: ServiceContext|undefined;
+  private stackdriverLog:
+      StackdriverLog;  // TODO: add type for @google-cloud/logging
+  constructor(options: Options) {
+    if (new.target !== LoggingBunyan) {
+      return new LoggingBunyan(options);
+    }
+    options = options || {};
+    super({objectMode: true});
+    this.logName = options.logName || 'bunyan_log';
+    this.resource = options.resource;
+    this.serviceContext = options.serviceContext;
+    this.stackdriverLog = logging(options).log(this.logName, {
+      removeCircular: true,
+    });
+  }
+
+  /**
+   * @param {string|number} level A bunyan logging level. Log entries at or
+   * above this level will be routed to Stackdriver Logging.
+   *
+   * @example
+   * const logger = bunyan.createLogger({
+   *   name: 'my-service',
+   *   streams: [
+   *     loggingBunyan.stream('info')
+   *   ]
+   */
+  stream(level: string|
+         number): {level: string|number, type: string, stream: {}} {
```

Comment:
It would be nice to see complex types used in public APIs pulled out to interfaces

Reply:
Done.